---
title: "Transfer labels from RC17"
author: "Nadine Bestard"
date: "`r Sys.Date()`"
output: 
  html_document:
    code_folding: hide
---

## Set up


```{r set-up, message=FALSE, warning=FALSE}
library(scRNAseq) #load laManno Dataset hES
library(SingleR) # annotate dataset
library(scuttle) # normalise reference
library(here) # reproducible paths
library(pheatmap) # diagnostic plots
library(scater) # dimplots
```

```{r}
project <- "TK_Mshef7"
source(here("src","R", "colours.R"))
```

```{r load-sce}
if (!(file.exists(
  here("processed", project,  "sce_rc17labels_into_mshef7.RDS")
))) {
sce <- readRDS(here("processed", project,  "sce_clusters_01.RDS"))
} else{
  sce <- readRDS(here("processed", project,  "sce_rc17labels_into_mshef7.RDS"))
}
```

# Import reference 
```{r load-refs}
sce_rc17 <- readRDS(here("processed", "TK_RC17", "sce_anno_01.RDS"))
```


# Annotate


```{r reference}
if(!file.exists(here("processed", project, "rc17_into_mshef7.RDS"))){
# switch to ENESMBLE notation in the query dataset
#rownames(sce) <- rowData(sce)$ID
# We choose wilcox as "This is slower but more appropriate for single-cell data compared to the default marker detection algorithm"
# from singleR vignette
annotation <- SingleR(sce, ref=sce_rc17, labels= sce_rc17$cluster_shortname, de.method = "wilcox")
# come back to gene row names
#rownames(sce) <- rowData(sce)$symb_uniq
saveRDS(annotation, here("processed", project, "rc17_into_mshef7.RDS"))
}else{
  annotation <- readRDS(here("processed", project, "rc17_into_mshef7.RDS"))
}
```

# Diagnostics
```{r}
plotScoreHeatmap(annotation)
```

# Transfer back to query dataset. 
```{r warning=FALSE}
sce$transfered_rc17 <- as.factor(annotation$pruned.labels)
# I want the NA as a level, for later representation. 
sce$transfered_rc17 <- forcats::fct_na_value_to_level(sce$transfered_rc17, "Unknown")

#plots
plotReducedDim(sce, dimred= "UMAP", colour_by = "transfered_rc17") + scale_colour_manual(values=cols)

plotReducedDim(sce, colour_by = "transfered_rc17", dimred = "UMAP", point_size = 0.1, other_fields = "transfered_rc17" ) + facet_wrap(~transfered_rc17) + scale_colour_manual(values=cols)

plotReducedDim(sce, colour_by = "transfered_rc17", dimred = "UMAP", point_size = 0.1, other_fields = "originalexp_snn_res.0.7", text_by =  "transfered_rc17" ) + facet_wrap(~originalexp_snn_res.0.7) + scale_colour_manual(values=cols)

table(sce$originalexp_snn_res.0.7, sce$transfered_rc17)

plotReducedDim(sce, colour_by = "transfered_rc17", dimred = "UMAP", point_size = 0.1, other_fields = c("originalexp_snn_res.0.7","transfered_rc17") , text_by =  "transfered_rc17" ) + facet_grid(transfered_rc17~originalexp_snn_res.0.7) + scale_colour_manual(values=cols)

```


<!-- # Add annotation using igrabski algorithm -->

<!-- ```{r} -->
<!-- if (!(file.exists( -->
<!--   here("processed", project,  "sce_labels.RDS") -->
<!-- ))) { -->
<!--   source(here("src/R/scRNAseq-cell-type/cell_type_identification4_clean.R")) -->

<!--   # train data -->
<!--   igrabski <- trainAllReference(counts(lamanno_es),lamanno_es$Cell_type) -->
<!--   # switch to ENESMBLE notation in the query dataset -->
<!--   rownames(sce) <- rowData(sce)$ID -->
<!--   annotation_igrabski <- classifyTarget(as.matrix(counts(sce)), igrabski, other = TRUE ) -->
<!--   sce$lamanno_es_ig <- annotation_igrabski -->
<!--   # put back annotation -->
<!--   rownames(sce) <- rowData(sce)$symb_uniq -->
<!-- } -->
<!-- ``` -->

<!-- # Compare both annotations -->
<!-- ```{r} -->
<!-- pheatmap(table(sce$lamanno_es, sce$lamanno_es_ig), cluster_rows = F, cluster_cols = F, scale = "row") -->
<!-- pheatmap(table(sce$lamanno_es, sce$lamanno_es_ig), cluster_rows = F, cluster_cols = F, scale = "column") -->

<!-- ``` -->


# Save

```{r save}
if (!(file.exists(
  here("processed", project,  "sce_rc17labels_into_mshef7.RDS")
))) {
saveRDS(sce, here("processed", project,  "sce_rc17labels_into_mshef7.RDS") )
}
```

