---
title: "Transfer labels"
author: "Nadine Bestard"
date: "`r Sys.Date()`"
output: 
  html_document:
    code_folding: hide
---




## Set up


```{r set-up, message=FALSE, warning=FALSE}
library(scRNAseq) #load laManno Dataset hES
library(SingleR) # annotate dataset
library(scuttle) # normalise reference
library(here) # reproducible paths
library(pheatmap) # diagnostic plots
library(scater) # dimplots
```

```{r}
project <- "TK_RC17"
source(here("src","R", "colours.R"))
```

```{r load-sce}
if (!(file.exists(
  here("processed", project,  "sce_labels.RDS")
))) {
  sce <- readRDS(here("processed", project,  "sce_corrected.RDS"))
} else{
  sce <- readRDS(here("processed", project,  "sce_labels.RDS"))
}
```

# Import reference 

```{r load-refs}
if (!(file.exists(
  here("processed", project,  "sce_labels.RDS")
))) {
lamanno_ips <- LaMannoBrainData("human-ips", ensembl=TRUE)
lamanno_es <- LaMannoBrainData("human-es", ensembl=TRUE)
lamanno_in_vivo <-  LaMannoBrainData('human-embryo', ensembl=TRUE)

genes <- intersect(intersect(rownames(lamanno_ips), rownames(lamanno_es)), rownames(lamanno_in_vivo))
lamanno <- cbind(lamanno_ips[genes,],lamanno_es[genes,], lamanno_in_vivo[genes,])
}
```

```{r normalise}
if (!(file.exists(
  here("processed", project,  "sce_labels.RDS")
))) {
lamanno <- logNormCounts(lamanno)
}
```

# Annotate


```{r reference}
if(!file.exists(here("processed", project, "annotation.RDS"))){
# switch to ENESMBLE notation in the query dataset
rownames(sce) <- rowData(sce)$ID
# We choose wilcox as "This is slower but more appropriate for single-cell data compared to the default marker detection algorithm"
# from singleR vignette
annotation <- SingleR(sce, ref=lamanno, labels= lamanno$Cell_type, de.method = "wilcox")
# come back to gene row names
rownames(sce) <- rowData(sce)$symb_uniq
saveRDS(annotation, here("processed", project, "annotation.RDS"))
}else{
  annotation <- readRDS(here("processed", project, "annotation.RDS"))
}
```

# Diagnostics


```{r}
plotScoreHeatmap(annotation)
# only ips
ips_labs <-grep("^i", unique(annotation$labels), value = TRUE )
ips_cells <- rownames(annotation)[annotation$labels %in% ips_labs]
plotScoreHeatmap(annotation, labels.use = ips_labs, cells.use = ips_cells)
#only es
es_labs <-grep("^e", unique(annotation$labels), value = TRUE )
es_cells <- rownames(annotation)[annotation$labels %in% es_labs]
plotScoreHeatmap(annotation, labels.use = es_labs, cells.use = es_cells)
#only in_vivo
in_vivo_labs <-grep("^h", unique(annotation$labels), value = TRUE )
in_vivo_cells <- rownames(annotation)[annotation$labels %in% in_vivo_labs]
plotScoreHeatmap(annotation, labels.use = in_vivo_labs, cells.use = in_vivo_cells)
```
# Transfer back to query dataset. 
```{r}
sce$lamanno <- as.factor(annotation$pruned.labels)
# I want the NA as a level, for later representation. 
levels(sce$lamanno) <-c(levels(sce$lamanno), "NA")
sce$lamanno[(is.na(sce$lamanno))] <- "NA"
#plots
plotReducedDim(sce, dimred= "UMAP", colour_by = "lamanno") + scale_colour_manual(values=cols)

plotReducedDim(sce, colour_by = "lamanno", dimred = "UMAP", point_size = 0.1, other_fields = "lamanno" ) + facet_wrap(~lamanno) + scale_colour_manual(values=cols)
```

# Add annotation from lammano es only
```{r}

if(!file.exists(here("processed", project, "annotation_es.RDS"))){
lamanno_es <- logNormCounts(lamanno_es)
# switch to ENESMBLE notation in the query dataset
rownames(sce) <- rowData(sce)$ID
# We choose wilcox as "This is slower but more appropriate for single-cell data compared to the default marker detection algorithm"
# from singleR vignette
annotation_es <- SingleR(sce, ref=lamanno_es, labels= lamanno_es$Cell_type, de.method = "wilcox")
# come back to gene row names
rownames(sce) <- rowData(sce)$symb_uniq
saveRDS(annotation_es, here("processed", project, "annotation_es.RDS"))

# transfer back
sce$lamanno_es <- as.factor(annotation_es$pruned.labels)
# I want the NA as a level, for later representation. 
levels(sce$lamanno_es) <-c(levels(sce$lamanno_es), "NA")
sce$lamanno_es[(is.na(sce$lamanno_es))] <- "NA"

}else{
  annotation_es <- readRDS(here("processed", project, "annotation_es.RDS"))
}
```

# Add annotation using igrabski algorithm

```{r}
if (!(file.exists(
  here("processed", project,  "sce_labels.RDS")
))) {
  source(here("src/R/scRNAseq-cell-type/cell_type_identification4_clean.R"))
  
  # train data
  igrabski <- trainAllReference(counts(lamanno_es),lamanno_es$Cell_type)
  # switch to ENESMBLE notation in the query dataset
  rownames(sce) <- rowData(sce)$ID
  annotation_igrabski <- classifyTarget(as.matrix(counts(sce)), igrabski, other = TRUE )
  sce$lamanno_es_ig <- annotation_igrabski
  # put back annotation
  rownames(sce) <- rowData(sce)$symb_uniq
}
```

# Add annotation using igrabski algorithm

```{r}
source(here("src/R/scRNAseq-cell-type/cell_type_identification4_clean.R"))

# train data
igrabski <- trainAllReference(counts(lamanno_es),lamanno_es$Cell_type)
# switch to ENESMBLE notation in the query dataset
rownames(sce) <- rowData(sce)$ID
annotation_igrabski <- classifyTarget(as.matrix(counts(sce)), igrabski, other = TRUE )
sce$lamanno_es_ig <- annotation_igrabski
# put back annotation
rownames(sce) <- rowData(sce)$symb_uniq
```

# Compare both annotations
```{r}
pheatmap(table(sce$lamanno_es, sce$lamanno_es_ig), cluster_rows = F, cluster_cols = F, scale = "row")
pheatmap(table(sce$lamanno_es, sce$lamanno_es_ig), cluster_rows = F, cluster_cols = F, scale = "column")

```


# A more global annotation

Using isabella grabski's algorithm again, but this time I will merge the similar clusters
in the training dataset.

```{r}
# group the cell types
Nb<- grep("^eNb", unique(lamanno_es$Cell_type), value = TRUE )
Rgl <- grep("^eRgl", unique(lamanno_es$Cell_type), value = TRUE )
Prog <- grep("^eProg", unique(lamanno_es$Cell_type), value = TRUE )
SC <- grep("^eSC", unique(lamanno_es$Cell_type), value = TRUE )

lamanno_es$grouped_celltype <- dplyr::case_when(lamanno_es$Cell_type %in% Nb ~ "Nb",
                                                lamanno_es$Cell_type %in% Rgl ~ "Rgl",
                                                lamanno_es$Cell_type %in% Prog ~ "Prog",
                                                lamanno_es$Cell_type %in% SC ~ "SC")

# train data with new classification
igrabski <- trainAllReference(counts(lamanno_es),lamanno_es$grouped_celltype)
# switch to ENESMBLE notation in the query dataset
rownames(sce) <- rowData(sce)$ID
annotation_igrabski_grouped <- classifyTarget(as.matrix(counts(sce)), igrabski, other = TRUE )
sce$celltype_es_ig <- annotation_igrabski_grouped
# put back annotation
rownames(sce) <- rowData(sce)$symb_uniq

```

### Compare with the other ig annotation
```{r}
pheatmap(table(sce$celltype_es_ig, sce$lamanno_es_ig), cluster_rows = F, cluster_cols = F, scale = "row")
pheatmap(table(sce$celltype_es_ig, sce$lamanno_es_ig), cluster_rows = F, cluster_cols = F, scale = "column")

```
# Add annotation from lammano in_vivo only
```{r}

if(!file.exists(here("processed", project, "annotation_in_vivo.RDS"))){
  
  lamanno_in_vivo <- logNormCounts(lamanno_in_vivo)
  # switch to ENESMBLE notation in the query dataset
  rownames(sce) <- rowData(sce)$ID
  # We choose wilcox as "This is slower but more appropriate for single-cell data compared to the default marker detection algorithm"
  # from singleR vignette
  annotation_in_vivo <- SingleR(sce, ref=lamanno_in_vivo, labels= lamanno_in_vivo$Cell_type, de.method = "wilcox")
  # come back to gene row names
  rownames(sce) <- rowData(sce)$symb_uniq
  saveRDS(annotation_in_vivo, here("processed", project, "annotation_in_vivo.RDS"))
  
  # transfer back
  sce$lamanno_in_vivo <- as.factor(annotation_in_vivo$pruned.labels)
  # I want the NA as a level, for later representation. 
  levels(sce$lamanno_in_vivo) <-c(levels(sce$lamanno_in_vivo), "NA")
  sce$lamanno_in_vivo[(is.na(sce$lamanno_in_vivo))] <- "NA"
  
}else{
  annotation_in_vivo <- readRDS(here("processed", project, "annotation_in_vivo.RDS"))
  # transfer back
  sce$lamanno_in_vivo <- as.factor(annotation_in_vivo$pruned.labels)
  # I want the NA as a level, for later representation. 
  levels(sce$lamanno_in_vivo) <-c(levels(sce$lamanno_in_vivo), "NA")
  sce$lamanno_in_vivo[(is.na(sce$lamanno_in_vivo))] <- "NA"
}
```


# Cell Cycle Annotation

```{r}
set.seed(10001)
hs_pairs <- readRDS(system.file("exdata", "human_cycle_markers.rds", package="scran"))
assigned <- cyclone(sce, pairs=hs_pairs, 
    gene.names=rowData(sce)$ID)
sce$phase <- assigned$phases
sce$phase_score_G1 <- assigned$scores$G1
sce$phase_score_G2M <- assigned$scores$G2M
sce$phase_score_S <- assigned$scores$S
saveRDS(assigned, here("processed",project,  "cyclone_asignments.RDS"))
```


# Save

```{r save}
if (!(file.exists(
  here("processed", project,  "sce_labels.RDS")
))) {
saveRDS(sce, here("processed", project,  "sce_labels.RDS") )
}
```

